Set github workflow permissions to least privileges

I would like to suggest setting the permissions to the github workflows as read only on the [top level](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#permissions) and any write permission be given at the [run level](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idpermissions).

This is necessary due to a behavior of github workflow to grant to GITHUB_TOKEN write permissions to all types of permissions, regardless of they being used or not. In case of the workflow getting compromised, an attacker can exploit this permissions.

This can be seen in the Action run step "Set up job" such as https://github.com/PyTables/PyTables/actions/runs/4407964071/jobs/7722340374:
 
<img width="670" alt="image" src="https://user-images.githubusercontent.com/22223372/225079761-130bcd1f-ab82-4d00-b580-a72de7f04fbd.png">


Thus, it is both a recommendation from [OpenSSF Scorecard](https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions) and the [Github](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions) to always use credentials that are minimally scoped.

Let me know if a PR is welcome and I'll submit it as soon as possible. 

Thanks!

Disclosure: I'm from Google, working with the OpenSSF to help many open source projects to increase their supply-chain security.

I won't only be able to test if it will be working with the new permissions, but looking at the code and considering that are mostly python script or make script that does not refer GITHUB_TOKEN, `contents: read` might be enough.

About the error on Wheels workflow bellow:
```
ERROR: Could not find a version that satisfies the requirement numexpr>=2.6.2 (from tables) (from versions: none)
ERROR: No matching distribution found for numexpr>=2.6.2
Error: Process completed with exit code 1.
```

When I tried to install the numexpr version 2.6.2 locally it also didn't work, although the error was quite different. Perhaps the version is broken now? 

<img width="478" alt="image" src="https://user-images.githubusercontent.com/22223372/225095645-20b15640-ec29-42d1-961c-7f53b55bc4e9.png">

but the version 2.8.4 installed successfully so the problem might be 2.6.2.

<img width="861" alt="image" src="https://user-images.githubusercontent.com/22223372/225095387-e83346e6-7b0a-484c-8f69-fa992935f43f.png">

You can consider upgrade the numexpr version and see if it works.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -2,6 +2,9 @@ name: CI
 
 on: [push, pull_request]
 
+permissions:
+  contents: read
+
 jobs:
 
   sdist:

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -2,6 +2,9 @@ name: ubuntu
 
 on: [push, pull_request]
 
+permissions:
+  contents: read
+
 jobs:
   build:
     runs-on: ubuntu-22.04

--- a/.github/workflows/wheels.yml
+++ b/.github/workflows/wheels.yml
@@ -13,6 +13,9 @@ on:
     types:
       - published
 
+permissions:
+  contents: read
+
 jobs:
   build_sdist:
     name: Build source distribution

